app/javascript/mastodon/stream.js

```diff
@@ -1,87 +1,191 @@
 import WebSocketClient from '@gamestdio/websocket';
 
-const randomIntUpTo = max => Math.floor(Math.random() * Math.floor(max));
+let sharedConnection;
+
+/**
+ * @typedef Subscription
+ * @property {string} channelName
+ * @property {object} params
+ * @property {function} onConnect
+ * @property {function} onReceive
+ * @property {function} onDisconnect
+ */
+
+const subscriptions = [];
+
+/**
+ * @param {Subscription} subscription
+ */
+const addSubscription = subscription => {
+  subscriptions.push(subscription);
+};
 
-const knownEventTypes = [
-  'update',
-  'delete',
-  'notification',
-  'conversation',
-  'filters_changed',
-];
+/**
+ * @param {Subscription} subscription
+ */
+const removeSubscription = subscription => {
+  const index = subscriptions.indexOf(subscription);
+  subscriptions.splice(index, 1);
+};
 
-export function connectStream(path, pollingRefresh = null, callbacks = () => ({ onConnect() {}, onDisconnect() {}, onReceive() {} })) {
-  return (dispatch, getState) => {
-    const streamingAPIBaseURL = getState().getIn(['meta', 'streaming_api_base_url']);
-    const accessToken = getState().getIn(['meta', 'access_token']);
-    const { onConnect, onDisconnect, onReceive } = callbacks(dispatch, getState);
+/**
+ * @param {Subscription} subscription
+ */
+const subscribe = subscription => {
+  const { channelName, params, onConnect } = subscription;
+  sharedConnection.send(JSON.stringify({ type: 'subscribe', stream: channelName, ...params }));
+  onConnect();
+};
 
-    let polling = null;
+/**
+ * @param {Subscription} subscription
+ */
+const unsubscribe = subscription => {
+  const { channelName, params, onDisconnect } = subscription;
+  sharedConnection.send(JSON.stringify({ type: 'unsubscribe', stream: channelName, ...params }));
+  onDisconnect();
+};
 
-    const setupPolling = () => {
-      pollingRefresh(dispatch, () => {
-        polling = setTimeout(() => setupPolling(), 20000 + randomIntUpTo(20000));
-      });
-    };
+const sharedCallbacks = {
+  connected () {
+    subscriptions.forEach(subscription => subscribe(subscription));
+  },
 
-    const clearPolling = () => {
-      if (polling) {
-        clearTimeout(polling);
-        polling = null;
+  received (data) {
+    const { stream } = data;
+
+    subscriptions.filter(({ channelName, params }) => {
+      const streamChannelName = stream[0];
+
+      if (stream.length === 1) {
+        return channelName === streamChannelName;
       }
-    };
 
-    const subscription = getStream(streamingAPIBaseURL, accessToken, path, {
-      connected () {
-        if (pollingRefresh) {
-          clearPolling();
-        }
+      const streamIdentifier = stream[1];
 
-        onConnect();
-      },
+      if (['hashtag', 'hashtag:local'].includes(channelName)) {
+        return channelName === streamChannelName && params.tag === streamIdentifier;
```

Comment:
Doesn't that prevent having multiple hashtag columns open for the same “main” hashtag but different configs overall?

Reply:
I'm trying to understand what you mean... Let's say we call `connectStream` 3 times for the same hashtag but different columns... Only 1 stream will be used (and server-side, only one subscription will be made), but each listener for the 3 different calls will be added to `subscriptions`. Then when a message arrives (indeed, only once, because server-side there is only one subscription), this code will match the 3 different client-side subscriptions and call their listeners respectively. If one of them disconnects...

Okay, if one of them disconnects, it will immediately send an unsubscribe to the server, closing the 1 server-side subscription there was, thus preventing other client-side subscriptions from receiving new events. Is that what you meant? That wouldn't be on this line though.

Reply:
Nevermind, I thought the filtering was made server-side, but it's actually the front-end making as many subscriptions as needed and filtering when receiving toots. Would have to re-read it, but it seems fine.

Reply:
I think the last piece I need to add is a ref counter so one client-side unsubscribe doesn't server-side unsubscribe when multiple listeners for the same stream exist.